storm-client/src/jvm/org/apache/storm/windowing/WindowManager.java

```diff
@@ -111,14 +125,86 @@ public void add(Event<T> windowEvent) {
             LOG.debug("Got watermark event with ts {}", windowEvent.getTimestamp());
         }
         track(windowEvent);
-        compactWindow();
+        if (!stateful) {
+            compactWindow();
+        }
     }
 
     /**
      * The callback invoked by the trigger policy.
      */
     @Override
     public boolean onTrigger() {
+        return stateful ? doOnTriggerStateful() : doOnTrigger();
```

Comment:
Strategy is not straightforward without a lot of refactoring. Will move it into sub class to keep the stateful case separate.